config/secret.go

```diff
@@ -0,0 +1,196 @@
+		return "", fmt.Errorf("accessing secret failed: %v", err)
+	}
+
+	if len(s.resolvers) == 0 {
+		return string(secret), nil
+	}
+
+	replaceErrs := make([]string, 0)
+	newsecret := secretPattern.ReplaceAllFunc(secret, func(match []byte) []byte {
+		resolver, found := s.resolvers[string(match)]
+		if !found {
+			replaceErrs = append(replaceErrs, fmt.Sprintf("no resolver for %q", match))
+			return match
+		}
+		replacement, _, err := resolver()
+		if err != nil {
+			replaceErrs = append(replaceErrs, fmt.Sprintf("resolving %q failed: %v", match, err))
+			return match
+		}
+
+		return []byte(replacement)
+	})
+	if len(replaceErrs) > 0 {
+		return "", fmt.Errorf("replacing secrets failed: %s", strings.Join(replaceErrs, ";"))
+	}
+
+	return string(newsecret), nil
+}
+
+// GetUnlinked return the parts of the secret that is not yet linked to a resolver
+func (s *Secret) GetUnlinked() []string {
+	return s.unlinked
+}
+
+// Link used the given resolver map to link the secret parts to their
+// secret-store resolvers.
+func (s *Secret) Link(resolvers map[string]telegraf.ResolveFunc) error {
+	// Setup the resolver map
+	s.resolvers = make(map[string]telegraf.ResolveFunc)
+
+	secret, err := s.read()
+	if err != nil {
+		return fmt.Errorf("accessing secret failed: %v", err)
+	}
+
+	// Iterate through the parts and try to resolve them. For static parts
+	// we directly replace them, while for dynamic ones we store the resolver.
+	replaceErrs := make([]string, 0)
+	newsecret := secretPattern.ReplaceAllFunc(secret, func(match []byte) []byte {
+		resolver, found := resolvers[string(match)]
+		if !found {
+			replaceErrs = append(replaceErrs, fmt.Sprintf("unlinked part %q", match))
+			return match
+		}
+		replacement, dynamic, err := resolver()
+		if err != nil {
+			replaceErrs = append(replaceErrs, fmt.Sprintf("resolving %q failed: %v", match, err))
+			return match
+		}
+
+		// Replace static parts right away
+		if !dynamic {
+			return []byte(replacement)
+		}
+
+		// Keep the resolver for dynamic secrets
+		s.resolvers[string(match)] = resolver
+		return match
+	})
+	if len(replaceErrs) > 0 {
+		return fmt.Errorf("linking secrets failed: %s", strings.Join(replaceErrs, ";"))
+	}
+
+	// Store the secret if it has changed
+	if string(secret) != string(newsecret) {
+		s.enclave = memguard.NewEnclave(newsecret)
+	}
+
+	// All linked now
+	s.unlinked = nil
+
+	return nil
+}
+
+func splitLink(s string) (storeid string, key string) {
+	// There should _ALWAYS_ be two parts due to the regular expression match
+	parts := strings.SplitN(s[2:len(s)-1], ":", 2)
+	return parts[0], parts[1]
+}
+
+func (s *Secret) read() ([]byte, error) {
+	if s.enclave == nil {
+		return nil, errors.New("secret not initialized")
+	}
+
+	lockbuf, err := s.enclave.Open()
+	if err != nil {
+		return nil, fmt.Errorf("opening enclave failed: %v", err)
+	}
+	defer lockbuf.Destroy()
+
+	var buf bytes.Buffer
+	_, err = io.Copy(&buf, lockbuf.Reader())
+
+	return buf.Bytes(), err
```

Comment:
The reasoning was the following. The `read` function shown above tries to guarantee the 

>  After use, the LockedBuffer should be destroyed.

part (see `defer lockbuf.Destroy()`). It is used in two places, the `Link` part and the `Get` part. In both places, we need the unencrypted secret to either link it to the secret stores (or replace static parts) or to return it to the plugin for further use (e.g. to add it to a query or authentication call). Unfortunately, we cannot return the `lockbuf` Reader, as it is not valid after destruction. Returning it to the plugin or `Link` caller is not an option, as we cannot guarantee the `lockbuf` is actually destroyed, running into instance limitations.

This leaves us with copying the `read` code to the two calling functions (`Link` and `Get`). For both functions, this has the gain to safe a placement on the stack, but the `lockbuf` destruction might be delayed and we have code duplication. Anyway, if you prefer to have it done this way, I can change the code.

In all cases, we, at some point, need to make the unencrypted secret accessible to the plugin for use. As most of the secrets are embedded into authentication calls requiring the plain-text password as string, this is unavoidable and will always lead to places where the secret is stored in unprotected memory. The memguard protection just pushes this part a bit down the pipe at least _enabling _ plugins to handle secrets safely.

Reply:
> The reasoning was the following. The `read` function shown above tries to guarantee the
> 
> > After use, the LockedBuffer should be destroyed.
>
> part (see `defer lockbuf.Destroy()`). It is used in two places, the `Link` part and the `Get` part. In both places, we need the unencrypted secret to either link it to the secret stores (or replace static parts) or to return it to the plugin for further use (e.g. to add it to a query or authentication call). Unfortunately, we cannot return the `lockbuf` Reader, as it is not valid after destruction. Returning it to the plugin or `Link` caller is not an option, as we cannot guarantee the `lockbuf` is actually destroyed, running into instance limitations.

Yes, I saw this which was why I mentioned that perhaps there was a pattern you could employ that would return an implementation of an interface that would have a method that would both read the secret out and then destroy the `LockedBuffer`, which would satisfy this recommendation and make that method sorta 'one time use'. Note, I was spit-balling here.

> This leaves us with copying the `read` code to the two calling functions (`Link` and `Get`). For both functions, this has the gain to safe a placement on the stack, but the `lockbuf` destruction might be delayed and we have code duplication. Anyway, if you prefer to have it done this way, I can change the code.
> 
> In all cases, we, at some point, need to make the unencrypted secret accessible to the plugin for use. As most of the secrets are embedded into authentication calls requiring the plain-text password as string, this is unavoidable and will always lead to places where the secret is stored in unprotected memory. The memguard protection just pushes this part a bit down the pipe at least _enabling _ plugins to handle secrets safely.

Reading the code, I don't think this is on the stack since a [Buffer is a variable-sized buffer of bytes with Read and Write methods](https://pkg.go.dev/bytes#Buffer). Even if it were on the stack, while it would be safe from the GC, it could still be [dumpable](https://man7.org/linux/man-pages/man5/core.5.html) and [swapped](https://www.kernel.org/doc/html/latest/admin-guide/mm/concepts.html#anonymous-memory) (at least on Linux).

It's understood that the unencrypted secret must be made available to the plugin to use. What I was trying to think through was the lifetime of the unencrypted secret. `memguard`'s goals are to "reduce the likelihood of sensitive data being exposed when in memory" with the idea that you leave it in `memguard`'s enclave as often as is possible and only pull it out when absolutely needed.

The way that I thought about this wrt `telegraf` was that once the secret was fetched from secure storage and placed under `memguard`, the plugins would ask `memguard` to decrypt the secret *every time they needed it and then immediately discard it* which could have a meaningful improvement of security *iff the plugins handled the decrypted secret safely* (which might include a [guarded allocation](https://github.com/awnumar/memguard/blob/master/examples/stream/stream.go#L42) but perhaps simply on the stack (to avoid GC; perhaps overwriting with 0s as part of a `defer` could be interesting as well...)). This is a meaningful improvement since the unencrypted secret is rarely in memory (in terms of CPU time) and therefore far less likely to be leaked unencrypted to disk. It is not perfect since there might be brief moments where it is exposed and of course if a debugger is attached (though, on Linux [`PR_SET_DUMPABLE`](https://man7.org/linux/man-pages/man2/prctl.2.html) is an option to prevent `ptrace`).

Looking at the code:
* `plugins/inputs/http/http.go:gatherURL()` fetches the username and password from the secure enclave when the URL is constructed, and it looks like [it is constructed each time `telegraf` runs the plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/http/http.go#L71) (ie, the URL (and therefore the username and password) doesn't seem to be cached; please correct me. This is good!). However, `plugins/inputs/http/http.go:gatherURL()` holds onto the memory for the rest of the function instead of wiping it, etc after `h.client.Do(request)`. This is a missed opportunity, not to mention the allocation is on the heap (see above)
* `plugins/inputs/sql/sql.go:Init()` fetches the DSN from the secure enclave and doesn't cache it (good!) and then holds onto it for the remainder of the function. This is a missed opportunity, not to mention the allocation is on the heap (see above)
* `plugins/inputs/sql/sql.go:Start()` fetches the DSN from the secure enclave and then holds onto it via `s.db, err = dbsql.Open(s.driverName, dsn)` until `plugins/inputs/sql/sql.go:Stop()` is called (I didn't chase down if `s.db.Close()` (called from `Stop()`) let go of it; presumably it would). It appears that `Start()` and `Stop()` are called around `s.startGathering()` (based on `plugins/common/shim/input.go`); I didn't confirm this further. The DSN is in memory for quite some time with this plugin. If I'm reading correctly, for the entire time that all input plugins are running (please correct me; this was something I inferred)
* `config/secret.go:Link()` fetches the secret and holds onto it for long enough to check to see if the secret currently stored in `memguard` is the same as the one fetched from secret storage (this is good). The allocation is on the heap (see above) and therefore could be improved

The `http` plugin, `sql:Init()` and `Link()` are quite close to what I think is desired in that they are fetching when they need it, though I think there are improvements on the timing of when they 'let go' of the unencrypted secret and more importantly, that (I believe) the allocation is on the heap and will linger in the GC. Ensuring it is on the stack, scribbling over it with 0s/non-blocking random data immediately after use or (gasp) using CGO with `calloc` (then scribble 0s/non-blocking random data and `free` after use) might be options.

The architecture for `Start()` input plugins that need to start, gather all the inputs, then `Stop()` all the input plugins that need to stop is a bit problematic and is at odds with how we'd like to use `memguard`. I'm not sure what to suggest here (`Start()`ing, gathering and `Stop()`ing all the secret storage-using input plugins *one at a time* and then doing all the non-secret storage-using input plugins like today could be a way forward; there are likely other choices). If nothing is done with this plugin, it should probably at a minimum be called out in documentation as a limitation of this input plugin's use of secret storage.

Reply:
First of all I moved the `read()` code into the two places where it was called and reduced the number of copies flying around. I cannot think of a way to reduce this further in those functions.
Regarding an interface function, I would rather not want that. While it might slightly extend the range of the encrypted secret it has multiple drawbacks making this very fragile as we give-up the control over the lockbuf

1. Only a limited number of `lockbuf` objects can exist at the same time (see [documentation](https://pkg.go.dev/github.com/awnumar/memguard#LockedBuffer)), so if the receiver holds a reference for too long we might run out of buffers.
2. As we do not have control over the potential implementation, we cannot guarantee that `lockbuf.Destroy()` is actually called, e.g. if the secret is never read.
3. A plugin can still copy the content of the lockbuf, reducing the overall gain at the cost of a higher complexity.

Regarding the plugins, I really appreciate your analysis and we should definitively look into those. However, I think the current PR cannot and should not try to solve all problems at once. Usually the plugins will call external function which usually do not make any attempt to secure credentials (database drivers, API implementations, etc). So let's push this down to the plugin level for now and *allow* for plugins which handle stuff securely, which is already a leap to the current state.